### Functional Programming in Python
In class, we briefly explored the Functional Programming in Python through lambda functions, map, filter, iterators, generators, and deocrators. This note will review those ideas. 

#### Lambda Functions 
The `lambda` keyword creates anonymous functions within Python. Programmers usually use `lambda` out of convenience because they don't require explicitly defining a new function. This is especially useful when the function you are writing won't be used again outside a particular context. `lambda`s usually serve as an input function to higher order functions (that require functional input), i.e. `map`, `filter`, `pandas.apply`, and others alike.

In [5]:
# Before compiling the following code snippets, write down what
# each individual lambda function will return in an inline comment. 
# If you think it returns an error, why would it be the case. 
 
    
(lambda val: val** 2)(2)
# 4
(lambda x, y: x + y)(2, 3)
# 5
#(lambda x, y: x + y)((2, 3))
# error
ga = (lambda s: s if 'General' in s else 'Specific ' + s)
ga('General Assembly') # General Assembly
ga('Assembly') # Specific ASsembly 

'Specific Assembly'

### Comprehensions 

In Python, you can build sequences using list comprehensions or generator expressions. Such expressions make your code more readable and often faster to execute. 

**Exercise**: Transform the following piece of code into a list comprehension.

    letters = string.ascii_uppercase
    letters_idx = []
    for letter in letters:
        letters_idx.append(letters.index(letter))
    print letters_idx
    

**Exercise**: Create a Cartesian product of t-shirt colors/sizes using a list comprehension. 
    
    Inputs: 
    colors = ['black', 'white']
    sizes = ['S', 'M', 'L']
    
    Output: 
    [('black', 'S'), ('black', 'M'), ('black', 'L'), 
     ('white', 'S'), ('white', 'M'), ('white', 'L')]
     
     
**Exercise:** In the output above, change the 'S' size to 'Small' and 'M' and 'L' to be 'Large'. 

**Exercise:** What are the other types of comprehensions and how are they different from list comprehensions?

While `for loop` can be general and do things beyond just outputting a list, list comprehensions are designed to build a new list. Don’t use that syntax if you are not doing something with the produced list. Additionally, if the list comprehension spans more than two lines, it is probably best to break it apart or rewrite as a plain old `for loop`.

In [7]:
import string
letters = string.ascii_uppercase

In [12]:
[letters.index(letter) for letter in letters]

[0,
 1,
 2,
 3,
 4,
 5,
 6,
 7,
 8,
 9,
 10,
 11,
 12,
 13,
 14,
 15,
 16,
 17,
 18,
 19,
 20,
 21,
 22,
 23,
 24,
 25]

In [16]:
colors = ['black', 'white']
sizes = ['S', 'M', 'L']
result = [(color,size) for color in colors for size in sizes]

In [19]:
def size_map(tup):
    color = tup[0]
    size = tup[1]
    if  size == 'S':
        size = 'Small'
    elif size == 'M':
        size = 'Medium'
    else:
        size =  'Large'
    return (color, size)
    
map(size_map,result)

[('black', 'Small'),
 ('black', 'Medium'),
 ('black', 'Large'),
 ('white', 'Small'),
 ('white', 'Medium'),
 ('white', 'Large')]

### Higher Order functions
Higher order functions take in function as an argument OR returns a function as result is a higher order
function. Some of the examples we have discussed so far are `map`, `sorted`, `filter`. 

`map(func, iterable)` applies the function over elements of an iterable. If you have multiple iterables, then `map(func, zip(iterA, iterB, iterC))` where your function takes in as many arguments 

From module: `functools`, recall the function `reduce`. `functools.reduce(function, iterable[, initializer])`. 

**Exercise**: Read through the Blog Post on MapReduce (http://michaelnielsen.org/blog/write-your-first-mapreduce-program-in-20-minutes/) [no need to implement] and explain how it's related to the `map` and `reduce` functions we have covered. 

`filter(pred, iterable)` applies a function that returns a bool value to each item in the iterable and returns only the items for which this is true. 



`sorted(iterable, key=func)` takes in an optional key argument that allows you to provide a function to be applied to each item for sorting. 

Beyond `sorted`, `max(seq)`, `min(seq)`, and `seq.sort()` also use keys to determine the values used for ordering elements in a sequence.


**Exercise**: What do the following uses of `sorted` return?
    
    # first example 
    fruits = ['strawberry','fig','apple','cherry','raspberry','banana']
    sorted(fruits, key=len)
    
    # second example
    sorted(fruits, key=lambda word: word[::-1])
    
   
    
**Exercise**: Write a function to return the two words with the highest alphanumeric score of uppercase letters:

    def alpha_score(upper_letters):
         """
         Computers the alphanumeric sum of letters in a
         string. Prerequisite: upper_letters is composed
         entirely of capital letters.
         
         """
        return sum(map(lambda l: 1 + ord(l) - ord('A'), upper_letters))

    
    alpha_score('ABC')  # => 6 = 1 ('A') + 2 ('B') + 3 ('C')

    
    def two_best(words):
        pass  # Your implementation here

    two_best(['hEllO', 'wOrLD', 'i', 'aM', 'PyThOn'])
  

In [59]:
fruits = ['strawberry','fig','apple','cherry','raspberry','banana']
sorted(fruits, key=lambda word: word[::-1])

['banana', 'apple', 'fig', 'raspberry', 'strawberry', 'cherry']

In [55]:
def alpha_score(upper_letters):
    return sum(map(lambda l: 1 + ord(l) - ord('A'), upper_letters))


#alpha_score('ABC')  # => 6 = 1 ('A') + 2 ('B') + 3 ('C')


def two_best(words):
    words = map(string.upper,words) # convert to upper case
    scores = map(alpha_score, words)
    return sorted(words, key=alpha_score, reverse=True)[0:2]


In [56]:
two_best(['hEllO', 'wOrLD', 'i', 'aM', 'PyThOn'])

['PYTHON', 'WORLD']

### Iterables vs. Generators 

Iterator object represents a stream of data returned one value at a time. Meanwhile, generators lazily compute the values on the fly without holding the contents of the list in place in memory. 


[Expert] **Exercise**: For each of the following scenarios, discuss whether it would be more appropriate to use a generator expression or a list comprehension:
- Searching for a given entity in the entries of a 1TB database.
    - Generator, because you don't want to keep the contents of the 1 TB database in memory. 
- Calculate cheap airfare using origin-to-destination flight information.
    - List comprehension, because you need to calculate all the options before you can compare them to each other.
- Finding the first palindromic Fibonacci number greater than 1,000,000.
    - Generator, because you only want to calculate each iteration until you reach one that satisfies the condition. So you want to calculate as few iterations as possible (in other words you want to be "lazy")
- Determine all multi-word anagrams of user-supplied 1000-character-or-more strings (very expensive to do).
    - Not sure. Generator?? 
- Return a list of all startups within 50 miles of San Francisco. 
    - Iterator, because we need a list of ALL the startups, so no point in getting them one at a time with a generator. 



[Challenge] **Exercise**: In class, we dicussed how to generate primes using the following function. 

    def primes_under(n):
    tests = []
    for i in range(2, n):
        if not any(map(lambda test: test(i), tests)):
            tests.append(make_divisibility_test(i))
            yield i

Change this function to generate composites. What is the 1000th composite number?

In [4]:
def make_divisibility_test(n):
    def divisible_by_n(m):
        return m % n == 0
    return divisible_by_n

div_by_3 = make_divisibility_test(3)
filter(div_by_3, range(10)) # generates 0, 3, 6, 9
make_divisibility_test(5)(10) # => True

True

In [48]:
def composites_under(num):
    tests = []
    for i in range(2, num):
        tests.append(make_divisibility_test(i))
        if any(map(lambda test: test(i), tests[:len(tests)-1])):
            yield i

In [49]:
composites_under_10000 = composites_under(10000)

In [44]:
next(composites_under_10000)

14

In [50]:
for i,num in enumerate(range(1000)):
    result = next(composites_under_10000)
    #print(str(i) + 'th composite: ' + str(next(composites_under)))
print result

1197


**Answer** = 1197

### Decorators 

Functional decorators are essentially syntactic sugar that help you anotate and simplify your code. While decorators may seem abstract right now, they pop up all over the place when reading large codebases. Thus, it's important to have an understanding of what they do and why they could be useful. 

As mentioned in class, decorators take another function as an argument. They may perform some processing with the decorated function and return it or replace it with another function. More concretely, the following two snippets of code return the same output: 

    @decorate
    def target():
        print 'running target()'
     
    def target():
        print running target()
    
    target = decorate(target)


Deocrators help reduce the repetition in your code. While repetition may seem harmless, it can introduce subtle bugs into your system. Take the use case of an e-commerce app that keeps track of the promos/coupons being offered on the site. As the company introduces new types of promotional deals, it will probably write new functions to calculate the discount % for that customer associated with the promo. However, we also want to maintain a master list of all of the promos to keep track of the `best_promo`.

In [34]:
# in the snippet below 
promos = []
def promotion(promo_func):
    promos.append(promo_func)
    return promo_func

@promotion
def fidelity(order):
    """5% discount for customers with 1000 or more fidelity points"""
    return order.total() * .05 if order.customer.fidelity >= 1000 else 0

@promotion
def bulk_item(order):
    """10% discount for each LineItem with 20 or more units"""
    discount = 0
    for item in order.cart:
        if item.quantity >= 20:
            discount += item.total() * .1
    return discount


@promotion
def large_order(order):
    """ 7% discount for orders with 10 or more distinct items"""
    distinct_items = {item.product for item in order.cart}
    if len(distinct_items) >= 10:
        return order.total() * .07
    return 0

def best_promo(order):
    """
    Select best discount available

    """
    return max(promo(order) for promo in promos)

In the code above, `promos` list starts out empty. `promotion` decorator returns `promo_func` unchanged but appends it to the `promos` list. Any function decorated by @promotion will be added to promos. This thus becomes an easy way to add items to the master promos list as new promotions are going live. 

**Exercise**: Familiarize yourself with the ipyparallel (http://people.duke.edu/~ccc14/sta-663-2016/19C_IPyParallel.html). Elaborate on why @remote and @parallel are useful decorators and specifically why could they come in handy when you are implementing ML algorithms that are easily parallelizable. 

[Expert] **Exercise**: Explore how decorators are used in Flask through these bloog posts (http://flask.pocoo.org/docs/0.11/quickstart/; http://flask.pocoo.org/docs/0.11/patterns/viewdecorators/). Describe your findings.

[Challenge] **Exercise**: Automatic Caching

Write the `cache`  decorator that automatically caches any calls to the decorated function. You can assume that all arguments passed to the decorated function will always be hashable types. Can you think of reasons why such a decorator would be useful?

Note: This is functionality is implemented as a decorator with functools.lru_cache

In [53]:
results = []
def cache(function):
    results.append(function)
    return function

@cache
def add_one(n):
    return n + 1


In [57]:
add_one(10)

11

In [55]:
results

[<function __main__.add_one>]

In [4]:
def cache(function):
    #results.append(function)
#     pass  # Your implementation here

@cache
def fib(n):
    return fib(n-1) + fib(n-2) if n > 2 else 1

fib(10)  # 55 (takes a moment to execute)
fib(10)  # 55 (returns immediately)
fib(100) # doesn't take forever
fib(400) # doesn't raise RuntimeError

IndentationError: expected an indented block (<ipython-input-4-cb3f3d525b5a>, line 5)

Notes from:
https://realpython.com/blog/python/primer-on-python-decorators/

In [2]:
def parent():
    print("Printing from the parent() function.")

    def first_child():
        return "Printing from the first_child() function."

    def second_child():
        return "Printing from the second_child() function."

    print(first_child())
    print(second_child())

In [3]:
parent()

Printing from the parent() function.
Printing from the first_child() function.
Printing from the second_child() function.


In [10]:
import time


def timing_function(some_function):

    """
    Outputs the time a function takes
    to execute.
    """

    def wrapper():
        t1 = time.time()
        some_function()
        t2 = time.time()
        return "Time it took to run the function: " + str((t2 - t1)) + "\n"
    return wrapper

@timing_function
def num_list():
    num_list = []
    for num in (range(0, 10000)):
        num_list.append(num)
    print("\nSum of all the numbers: " + str((sum(num_list))))
    
# replace this...
# num_list = timing_function(num_list)

# with this...
num_list() # you're actually calling the wrapper() function, with some_function = num_list()
# we say that @timing_function decorates num_list


Sum of all the numbers: 49995000


'Time it took to run the function: 0.00416994094849\n'

In [9]:
wrapped_num_list()


Sum of all the numbers: 49995000


'Time it took to run the function: 0.006432056427\n'

In [11]:
from time import sleep


def sleep_decorator(function):

    """
    Limits how fast the function is
    called.
    """

    def wrapper(*args, **kwargs):
        sleep(2)
        return function(*args, **kwargs)
    return wrapper


@sleep_decorator
def print_number(num):
    return num

print(print_number(222))

for num in range(1, 6):
    print(print_number(num))

222
1
2
3
4
5
